File: Data Structures & Algorithms/decode-ways/submission-2.cpp
class Solution {
public:
    int solve(int index,string& s,auto& dp){
        if(index >= s.size())return 1;
        if(dp[index] != -1)return dp[index];
        // single char
        int ans = 0;
        if(s[index] >= '1' && s[index] <= '9'){
            ans = solve(index + 1, s, dp);
            if(index < s.size()-1){
                if(s[index] == '1' && s[index+1] >= '0' && s[index+1] <= '9'){
                    ans += solve(index+2,s, dp);
                }
                if(s[index] == '2' && s[index+1] >= '0' && s[index+1] <= '6'){
                    ans+= solve(index+2,s ,dp);
                }
            }
        }
        return dp[index] = ans;
    }   
    int numDecodings(string s) { 
        int a = 0, b = 1 , c = 1;
        for(int index = s.size()-1; index >= 0 ; index--){
            a = 0;
            if(s[index] >= '1' && s[index] <= '9'){
                a = b;
                if(index < s.size()-1){
                    if(s[index] == '1' && s[index+1] >= '0' && s[index+1] <= '9'){
                        a += c;
                    }
                    if(s[index] == '2' && s[index+1] >= '0' && s[index+1] <= '6'){
                        a+= c;
                    }
                }
            }
            c= b;
            b = a;
        }
        return a;
    }
};

File: Data Structures & Algorithms/coin-change/submission-2.java
class Solution {
    int[][] dp;
    private int solve(int[] coins,int amount,int index){
        if(amount == 0)return 0;
        if(index == coins.length)return 1000000009;
        if(dp[index][amount] != -1)return dp[index][amount];
        int notTake = solve(coins,amount,index+1);
        int take  = amount >= coins[index] ? 1+solve(coins,amount-coins[index],index) : 1000000009;
        return dp[index][amount] = Math.min(take,notTake);
    }
    public int coinChange(int[] coins, int amount) {
        dp = new int[coins.length][amount+1];
        for(int i=0;i<coins.length;i++){
            for(int j=0;j<=amount;j++){
                dp[i][j]=-1;
            }
        }
        int ans = solve(coins,amount,0);
        return ans == 1000000009 ? -1 : ans;
    }
}
